Implement Track-Cluster Subtraction (PFA1a)

ruse-traveler opened the pull request:
### Briefly, what does this PR introduce?

This PR implements the `TrackClusterSubtraction` algorithm, which ingests clusters from a specified calorimeter, any matched tracks, and subtracts off the expected energy deposited of the sum of tracks pointing to each cluster. This is the second stage of of the baseline particle flow routine, _PFAlpha_

This algorithm outputs two collections of clusters:
1. **Remnant clusters,** which consist of un-matched clusters and any cluster with nonzero energy left after the subtraction; and
2. **Expected clusters,** which consist of matched clusters with any remnant energy subtracted from them.

<img width="619" height="650" alt="PFA1_VisualExplainer" src="https://github.com/user-attachments/assets/9ed04da0-bc04-4f7a-83ea-7d8c5bdb7bbc" />

In both cases, all properties of the cluster before subtraction are copied to the remnant and expected clusters with energies updated accordingly. The remnant clusters will be used to reconstruct neutral particles downstream, and the expected clusters will be used in the kinematic reconstruction of charged particles downstream.

Users are provided two methods to check if the leftover energy is consistent with 0:
1. A **basic method,** which checks if the leftover energy is less than 0 or within floating point precision of 0; or a
2. **Resolution method,** which checks if the leftover energy is less than 0 or within some number of standard deviations of 0 based on the resolution of the tracker and calorimeter.

For the latter method, the number of sigma to use and the resolutions of the tracker/calorimeter are configurable parameters. 

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #1341 #2127)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No.

### Does this PR change default behavior?
No.

### To-do
Will follow up with a benchmark for this stage of PFAlpha with a benchmark.

ruse-traveler committed:
Add skeletons of algorithm and factory

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Propagate upstream changes from development fork

ruse-traveler committed:
Merge branch 'add-track-cluster-subtraction-pfa-one' of github.com:eic/EICrecon into add-track-cluster-subtraction-pfa-one

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Load particle plugin

ruse-traveler committed:
Use track PDGs for masses

ruse-traveler committed:
Fix merge conflicts

ruse-traveler committed:
Allow for checking zero against resolutions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Write out all relevant collections

ruse-traveler committed:
Fix merge conflicts

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Add particle category to global reconstruction

ruse-traveler committed:
Merge branch 'add-track-cluster-subtraction-pfa-one' of github.com:eic/EICrecon into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Merge main

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wdconinc committed:
Implement Track-Cluster Subtraction (PFA1) (fix: iwyu) (#2051)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/17247257038.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Remove PFTools

ruse-traveler committed:
Make sure cmake finds plugin

ruse-traveler committed:
Use track-cluster matches until merge/splitter is updated

ruse-traveler committed:
Make consistent with reco CMakeLists

ruse-traveler committed:
Do light reformatting

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Fix merge conflicts

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Remove unused ChangeRun, make Process consistent with base class

ruse-traveler committed:
Add outputs to defaults

ruse-traveler committed:
Clean up, add header guards

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wdconinc committed:
Implement Track-Cluster Subtraction (PFA1) (fix: iwyu) (#2087)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/17839250702.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Add FHCal insert

ruse-traveler committed:
Make sure FHCal insert collections are in default output

ruse-traveler committed:
Rename factories after primary output

ruse-traveler committed:
Remove WIP notes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Remove edm4eic 8.X guards

ruse-traveler committed:
Fix typo in docstring

wdconinc committed:
Implement Track-Cluster Subtraction (PFA1a) (fix: iwyu) (#2133)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/18477471601.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Remove unnecessary header

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wdconinc commented on src/algorithms/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@ cmake_minimum_required(VERSION 3.16)
 add_subdirectory(interfaces)
 add_subdirectory(calorimetry)
 add_subdirectory(tracking)
+add_subdirectory(particle)
```
For consideration, isn't `particle` a bit too generic and might we use `particle_flow` here?

wdconinc commented on src/algorithms/particle/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+plugin_add_algorithms(${PLUGIN_NAME})
+plugin_add_dd4hep(${PLUGIN_NAME})
+plugin_add_event_model(${PLUGIN_NAME})
+plugin_add_eigen3(${PLUGIN_NAME})
```
Appears to be unused.

wdconinc commented on src/global/particle/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+cmake_minimum_required(VERSION 3.16)
```
Inherit from top level.

wdconinc commented on src/algorithms/particle/TrackClusterSubtractorConfig.h:
```diff
@@ -0,0 +1,33 @@
+struct TrackClusterSubtractorConfig {
+
+  ///! fraction of track energy to subtract
+  double fracEnergyToSub = 1.0;
```
Don't use abbreviations; it affects clarity.

wdconinc commented on src/algorithms/particle/TrackClusterSubtractorConfig.h:
```diff
@@ -0,0 +1,33 @@
+  ///! fraction of track energy to subtract
+  double fracEnergyToSub = 1.0;
+
+  ///! default mass to use for track energy
```
Not mass but pdg.

wdconinc commented on src/algorithms/particle/TrackClusterSubtractorConfig.h:
```diff
@@ -0,0 +1,33 @@
+  int32_t defaultMassPdg = 211;
+
+  ///! index of surface to use for measuring momentum
+  uint64_t surfaceToUse = 1;
```
This is poorly defined; where do these surfaces come from and do they start counting from 0? Do we expect uint64_t many surfaces?

ruse-traveler replied:
In the latest round of commits I reduced the `unit64_t` to a `unint8_t` and pointed out the algorithm! When testing out some stuff, I also confirmed that the projections start at 1 (e.g. in the attached screenshot)...

<img width="588" height="614" alt="SampleProjections" src="https://github.com/user-attachments/assets/c8e45a90-3149-400d-b7a9-b0f62d496cd1" />

wdconinc commented on src/algorithms/particle/TrackClusterSubtractorConfig.h:
```diff
@@ -0,0 +1,33 @@
+  uint32_t nSigmaMax = 1;
+
+  ///! tracking momentum resolution to use
+  double trkReso = 1.0;
```
Avoid abbreviations.

wdconinc commented on src/algorithms/particle/TrackClusterSubtractor.h:
```diff
@@ -0,0 +1,90 @@
+  };
+
+  ///! Alias for vectors of clusters
+  using VecClust = std::vector<edm4eic::Cluster>;
```
Avoid public types if only needed privately.

wdconinc commented on src/algorithms/particle/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,215 @@
+                                     const TrackClusterSubtractor::Output& output) const {
+
+  // grab inputs/outputs
+  const auto [in_match, in_cluster, in_project] = input;
```
Variable names plural for collections.

wdconinc commented on src/algorithms/particle/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,215 @@
+
+  // calculate nSigma
+  const double totReso =
+      std::sqrt((m_cfg.trkReso * m_cfg.trkReso) + (m_cfg.calReso * m_cfg.calReso));
```
You could work in variances instead of taking sqrt.

wdconinc commented on src/algorithms/particle/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,215 @@
+    // ------------------------------------------------------------------------
+    auto expect_clust = cluster.clone();
+    expect_clust.setEnergy(cluster.getEnergy() - eSubToUse);
+    out_expect->push_back(expect_clust);
```
Avoid clone and push_back; use create.

wdconinc commented on src/algorithms/particle/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  // --------------------------------------------------------------------------
+  // 1. Build map of clusters onto projections
```
This first step seems to indicate we are missing a collection of associations that should be determined ahead of this and would avoid a search.

ruse-traveler replied:
I think this would be a good direction to explore, but I'd also like to run it past the PWGs & DSCs since it would have implications for analyses... Would you be okay flagging this as a follow-up item?

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Rename particle to particle_flow

ruse-traveler committed:
Inherit minimum version from top-level

ruse-traveler committed:
Remove unnecessary plugins

ruse-traveler committed:
Use centralized ObjectID comparator

ruse-traveler committed:
Make remaining alias private

ruse-traveler committed:
Make comments more stylistically consistent, remove redundant ones

ruse-traveler committed:
Improve parameter/variable names

ruse-traveler committed:
Call particle service only once

ruse-traveler committed:
Avoid sqrt in nsigma calculation

ruse-traveler committed:
Write out track-cluster links if running w/ EDM4eic 8.7.0 or greater

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Link against datamodel

ruse-traveler committed:
Update PF plugin name

ruse-traveler committed:
Link against particle service & tracking

epic-capybara committed:
Implement Track-Cluster Subtraction (PFA1a) (fix: iwyu) (#2599)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/24102434127.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

mandrenguyen commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,229 @@
+
+    // check if consistent with zero,
+    // set eSub accordingly
+    const bool isZero             = is_zero(eSubtracted);
```
Apologies, if my comments are all quite high level. 
The function name `is_zero` hides a bit the fact that if the track "energy" is greater than that of the linked calo clusters it will treat them as if they are compatible. 

I believe you we are inevitably going to have tracks for which we don't want to directly use their kinematics "as is".  I imagine we are always going to want to store tracks with a reasonably high efficiency selection, admitting that some will be poorly measured.  Depending on the working point, you might choose to accept all tracks on the condition of calo compatibility (or not).  However, in the case when the we don't find enough calo energy, I think it's inevitable that one will want to apply a tightened set of track selection criteria to try to reject some fakes.  Of course some of them won't be fakes at all,  but rather muons. But I think that one nevertheless has to leverage the calo compatibility in particle flow against fake tracks.   

This may be more of a comment for the future if the track selection, detector conditions, etc.  is not sufficiently well known yet.   You might consider some place holders though.   Or maybe at this point just a renaming of this function to make it a bit more transparent, and a little comment to indicate what's happening. 

A related comment is that you can also partially handle this in the momentum assignment.  If you don't think the track parameters are trustworthy (e.g., few hits, bad chi2, etc.) you can choose to weight towards the calo measurement, compared to the one from the tracker.

ruse-traveler replied:
Ah! That's a great point! Truth be told, my focus here was more on getting machinery in to do track-cluster subtraction, so I hadn't given much thought to how we would handle fake tracks. But I agree the calo compatibility is a really important metric!

For now, I'd be happy to rename the function to `is_track_energy_greater_than_calo` (or similar),  add some comments, and flag this as a important direction for future development!

I wonder what this might look like in implementation? I could imagine an algorithm that filters tracks based on their compatibility, and the "real" tracks are passed downstream to be converted to particles. Alternatively, I could also see this happening in the particle conversion step...

ruse-traveler replied:
> A related comment is that you can also partially handle this in the momentum assignment. If you don't think the track parameters are trustworthy (e.g., few hits, bad chi2, etc.) you can choose to weight towards the calo measurement, compared to the one from the tracker.

Side note: we actually have something gesturing in this direction in PFA3 currently, where you can take the resolution-weighted average of the track and cluster when calculating the particle energy!

ruse-traveler replied:
Just pushed some changes renaming `is_zero` to be a little more descriptive of the actual check!

mandrenguyen commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,229 @@
+  if (m_cfg.doNSigmaCut) {
+
+    // calculate n sigma squared
+    const double resolution2 = (m_cfg.trackResolution * m_cfg.trackResolution) +
```
Are we treating trackResolution as a single external parameter?  Can't we grab the uncertainty directly from the Kalman fit?

ruse-traveler replied:
Ah! True! We do store the covariance matrix with the other track parameters, so we could pull the track uncertainty from there. The only catch is that we currently don't propagate any uncertainties on the calorimeter clusters, and I'm not sure how close we are to reasonably calculating them 🫤 

So the rationale behind treating the track resolution as an external parameter was mostly for simplicity, and also to treat it and the calorimeter resolution on similar footing...

ruse-traveler replied:
Just pushed some changes which remove `trackResolution` and instead use the covariance on the projection momentum directly!

ruse-traveler committed:
Divide projections up by theta

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Remove spurious plugin

ruse-traveler committed:
Merge branch 'add-track-cluster-subtraction-pfa-one' of github.com:eic/EICrecon into add-track-cluster-subtraction-pfa-one

epic-capybara committed:
Implement Track-Cluster Subtraction (PFA1a) (fix: iwyu) (#2638)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/25022347249.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Shrink surface parameter to uint8_t; point out algorithm in doxygen comment

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler commented on src/global/particle_flow/particle_flow.cc:
```diff
@@ -0,0 +1,159 @@
+      {130.0 * dd4hep::degree, 177.0 * dd4hep::degree}};
+
+  auto subDivideByTheta =
+      [&thetaRanges](const edm4eic::TrackSegment& projection) -> std::vector<int> {
```
```suggestion
      [thetaRanges](const edm4eic::TrackSegment& projection) -> std::vector<int> {
```

ruse-traveler committed:
Not capture theta ranges by reference

Avoid stack-after-use error by capturing theta ranges by _value_ not by reference.

Co-authored-by: Derek M Anderson <derek.murphy.anderson@protonmail.com>

ruse-traveler committed:
Merge branch 'add-track-cluster-subtraction-pfa-one' of github.com:eic/EICrecon into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Use track covariance directly rather than as an external parameter

ruse-traveler committed:
Clean up: update copyrights, make is_zero more descriptive, improve docstrings

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler commented on src/factories/particle_flow/particle_flow.cc:
```diff
@@ -0,0 +1,93 @@
+      {"EcalEndcapNTrackClusterMatches", "EcalEndcapNClusters", "CalorimeterTrackProjections"},
+      {"EcalEndcapNRemnantClusters", "EcalEndcapNExpectedClusters",
+       "EcalEndcapNTrackExpectedClusterMatches"},
+      {.fracEnergyToSub = 1.0, .defaultMassPdg = 211, .surfaceToUse = 1},
```
In the latest round of commits, I pushed some changes that filter the projections by theta to reduce some of the combinatorics!

ruse-traveler commented:
All right! I think this is ready for another round of reviewing! The biggest changes in the latest round of commits are:

1. Projections are now separated by theta ahead of the algorithm in an effort to mitigate combinatorics; and
2. I've removed the `trackResolution` parameter in favor of working directly with the projection covariance matrix.

I think the combinatorics can be further mitigated by introducing a link between projections and clusters, but I feel that that is a little out-of-scope of this PR.

And -- much further down the road -- it would be preferable to treat the clusters and tracks on equal footing and use the variances from both rather than treating the calorimeter resolution as an external parameter. But I think we're a long ways away from calculating/using the cluster variances...

veprbl commented:
```
In file included from /home/runner/work/EICrecon/EICrecon/src/global/particle_flow/particle_flow.cc:21:
/home/runner/work/EICrecon/EICrecon/src/factories/particle_flow/TrackClusterSubtractor_factory.h:47:76: error: no member named 'trackResolution' in 'eicrecon::TrackClusterSubtractorConfig'
   47 |   ParameterRef<double> m_trackResolution{this, "trackResolution", config().trackResolution};
      |                                                                   ~~~~~~~~ ^
1 error generated.
```

ruse-traveler commented on src/factories/particle_flow/TrackClusterSubtractor_factory.h:
```diff
@@ -0,0 +1,76 @@
+  ParameterRef<uint8_t> m_surfaceToUse{this, "surfaceToUse", config().surfaceToUse};
+  ParameterRef<bool> m_doNSigmaCut{this, "doNSigmaCut", config().doNSigmaCut};
+  ParameterRef<uint32_t> m_nSigmaMax{this, "nSigmaMax", config().nSigmaMax};
+  ParameterRef<double> m_trackResolution{this, "trackResolution", config().trackResolution};
```
```suggestion
```

ruse-traveler committed:
Remove track resolution parameter

ruse-traveler commented:
> ```
> In file included from /home/runner/work/EICrecon/EICrecon/src/global/particle_flow/particle_flow.cc:21:
> /home/runner/work/EICrecon/EICrecon/src/factories/particle_flow/TrackClusterSubtractor_factory.h:47:76: error: no member named 'trackResolution' in 'eicrecon::TrackClusterSubtractorConfig'
>    47 |   ParameterRef<double> m_trackResolution{this, "trackResolution", config().trackResolution};
>       |                                                                   ~~~~~~~~ ^
> 1 error generated.
> ```

😩 Whoops! Just fixed it...

epic-capybara committed:
Implement Track-Cluster Subtraction (PFA1a) (fix: iwyu) (#2644)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/25170954335.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Include vector utils header

ruse-traveler committed:
Fix link handling

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

epic-capybara committed:
Implement Track-Cluster Subtraction (PFA1a) (fix: iwyu) (#2663)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/25559460896.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

veprbl committed:
Delete src/factories/TrackClusterSubtractor_factory.h

veprbl commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,276 @@
+      denominator += point.momentum[i] * point.momentum[i];
+    }
+    return denominator < std::numeric_limits<double>::epsilon() ? 0.0 : numerator / denominator;
+  };
```
This is *not* how you calculate the variance of a Eucledian norm if you know vector itself and its covariance.

veprbl commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,276 @@
+        continue;
+      } else {
+        momentum         = edm4hep::utils::magnitude(point.momentum);
+        momentumVariance = getVariance(point);
```
Here it is assumed that momentum is a 3 vector, like it says in
https://github.com/eic/EDM4eic/blob/7d7705d9bbd4794b1b875c04258e7d986aea8a47/edm4eic.yaml#L166-L167
Well, that documentation lies to you, here is what it really is:
https://github.com/eic/EICrecon/blob/a48dc9371f3459fe3c04ec866b8a38b319cffd71/src/algorithms/tracking/TrackProjector.cc#L133-L136
Perhaps, you only want modulo of the first component.

ruse-traveler replied:
Following up on the discussion in the Reco WG yesterday, I've reverted the variance calculation back to the "dumb" approach of treating the resolution as a parameter. I've opened up a issue to flag this as an area to improve on:

https://github.com/eic/EICrecon/issues/2687

Echoing my thoughts from the meeting: this most likely isn't difficult to implement, but I don't trust myself to be able to work though it & validate it quick enough to make it in the release. And I think at this stage even the less sophisticated version is still a good step forward!

veprbl commented on src/global/particle_flow/particle_flow.cc:
```diff
@@ -24,6 +35,128 @@ void InitPlugin(JApplication* app) {
+      {"HcalBarrelTrackClusterMatches", "HcalBarrelClusters", "BarrelCalorimeterTrackProjections"},
+      {"HcalBarrelRemnantClusters", "HcalBarrelExpectedClusters",
+#if EDM4EIC_BUILD_VERSION >= EDM4EIC_VERSION(8, 7, 0)
+       "HcalBarrelTrackExpectedClusterLink",
```
```suggestion
       "HcalBarrelTrackExpectedClusterLinks",
```

veprbl commented on src/factories/particle_flow/TrackClusterSubtractor_factory.h:
```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2025 Derek Anderson
```
```suggestion
// Copyright (C) 2026 Derek Anderson
```

veprbl commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,276 @@
+#include <podio/RelationRange.h>
+#include <podio/detail/Link.h>
+#include <podio/detail/LinkCollectionImpl.h>
+#include <stdint.h>
```
```suggestion
#include <cstdint>
```
I hope, IWYU will be happy with that too.

veprbl commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,276 @@
+    const uint32_t nSigmaMax2 = m_cfg.nSigmaMax * m_cfg.nSigmaMax;
+
+    isGreaterThan = (nSigma2 < nSigmaMax2);
+    trace("Within {} NSigma^2, track energy sum greater than calorimeter cluster: difference = {} "
```
Did you want to have `if (isGreaterThan) {` here?

ruse-traveler replied:
Whoops! Thank you for catching that!

veprbl commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,276 @@
+    auto expect_cluster = cluster.clone();
+    expect_cluster.setEnergy(cluster.getEnergy() - eSubtractedToUse);
+    out_expectants->push_back(expect_cluster);
+    trace("Created subtracted cluster with {} GeV (originally {} GeV)", expect_cluster.getEnergy(),
```
```suggestion
    trace("Created expected cluster with {} GeV (originally {} GeV)", expect_cluster.getEnergy(),
```
Would be in agreement to messages below.

veprbl commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,276 @@
+  // --------------------------------------------------------------------------
+  // 4. Any unmatched clusters are remnants by definition
+  // --------------------------------------------------------------------------
+  for (const auto& cluster : vecNoMatchClust) {
```
There is no other use to vecNoMatchClust, why not merge this and previous loop?

ruse-traveler replied:
Ah! Good point!

veprbl commented on src/global/particle_flow/particle_flow.cc:
```diff
@@ -24,6 +35,128 @@ void InitPlugin(JApplication* app) {
+  std::vector<std::pair<double, double>> thetaRanges{
+      {3.0 * dd4hep::degree, 50.0 * dd4hep::degree},
+      {45.0 * dd4hep::degree, 135.0 * dd4hep::degree},
+      {130.0 * dd4hep::degree, 177.0 * dd4hep::degree}};
```
These numbers will require some tweaking later. We need to file an issue for that.

veprbl commented on src/global/particle_flow/particle_flow.cc:
```diff
@@ -24,6 +35,128 @@ void InitPlugin(JApplication* app) {
+      }
+    }
+    return binIndices;
+  };
```
Why subdivide by theta, should it not be by system IDs (encoded as "extra" in acts surface id's)?

ruse-traveler replied:
True! System IDs would also work here! I went with theta initially just because I was thinking to divide up the projections by hemisphere, and there was an already existing example of using theta with the `FilterMatching` algorithm.

We could also get the system IDs from the points, too. Maybe something like this?
```c++
auto subdivideBySystemID = [systemIDs](const edm4eic::TrackSegment& projection) -> std::vector<int> {
  const int id = projection.getPoints()[0].system
  std::vector<int> idIndices;
  for (std::size_t iID = 0; iID < systemIDs.size(); ++iID) {
    if (id == systemIDs[iID]) {
      idIndices.push_back(iID);
    }
  }
  return idIndices;
};
```

veprbl replied:
Looks good to me.

veprbl reviewed (commented):
Nice piece of functionality and is long overdue. I have some comments.

ruse-traveler committed:
Merge 'main' into branch

ruse-traveler committed:
Apply suggestions from code review

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

ruse-traveler committed:
Divide projections by system IDs

ruse-traveler committed:
Merge unmatched cluster loops

ruse-traveler committed:
Wrap debug statement in conditional

ruse-traveler committed:
Revert to using manually set resolution

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

epic-capybara committed:
Implement Track-Cluster Subtraction (PFA1a) (fix: iwyu) (#2688)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/26851115047.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Update src/global/particle_flow/particle_flow.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,237 @@
+    }
+  } else {
+    isGreaterThan = difference < std::numeric_limits<double>::epsilon();
+    trace("Track energy sum greater than calorimeter cluster: difference = {} GeV", difference);
```
```suggestion
    if (isGreaterThan) {
      trace("Track energy sum greater than calorimeter cluster: difference = {} GeV", difference);
    }
```

veprbl commented on src/global/particle_flow/particle_flow.cc:
```diff
@@ -101,6 +110,139 @@ void InitPlugin(JApplication* app) {
+
+  app->Add(new JOmniFactoryGeneratorT<TrackClusterSubtractor_factory>(
+      "HcalEndcapPInsertRemnantClusters",
+      {"HcalEndcapPInsertTrackSplitMergeClusterMatches", "HcalEndcapPInsertClusters",
```
```

Error: JEventProcessorPODIO] [error] Omitting PODIO collection 'HcalEndcapPInsertExpectedClusters' due to exception: JOmniFactory: Failed to get collection HcalEndcapPInsertTrackSplitMergeClusterMatches: Missing databundle with uniquename 'HcalEndcapPInsertTrackSplitMergeClusterMatches'.
Error: JEventProcessorPODIO] [error] Omitting PODIO collection 'HcalEndcapPInsertRemnantClusters' due to exception: JOmniFactory: Failed to get collection HcalEndcapPInsertTrackSplitMergeClusterMatches: Missing databundle with uniquename 'HcalEndcapPInsertTrackSplitMergeClusterMatches'.
Error: JEventProcessorPODIO] [error] Omitting PODIO collection 'HcalEndcapPInsertTrackExpectedClusterLinks' due to exception: JOmniFactory: Failed to get collection HcalEndcapPInsertTrackSplitMergeClusterMatches: Missing databundle with uniquename 'HcalEndcapPInsertTrackSplitMergeClusterMatches'.
Error: JEventProcessorPODIO] [error] Omitting PODIO collection 'HcalEndcapPInsertTrackExpectedClusterMatches' due to exception: JOmniFactory: Failed to get collection HcalEndcapPInsertTrackSplitMergeClusterMatches: Missing databundle with uniquename 'HcalEndcapPInsertTrackSplitMergeClusterMatches'.
```
https://github.com/eic/EICrecon/actions/runs/26891856212/job/79320942404#step:7:643

ruse-traveler replied:
Whoops! Forgot that we don't pass the insert matches through the split-merger yet 😅

veprbl commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,237 @@
+        momentum = edm4hep::utils::magnitude(point.momentum);
+        break;
+      }
+    }
```
```suggestion
    double momentum = 0.0;
    bool momentum_valid = false;
    for (const auto& point : project.getPoints()) {
      if (point.surface != m_cfg.surfaceToUse) {
        continue;
      } else {
        momentum = edm4hep::utils::magnitude(point.momentum);
        momentum_valid = true;
        break;
      }
    }
    if (!momentum_valid) {
      continue;
    }
```

ruse-traveler committed:
Apply suggestions from code review

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Use correct FHCal insert track-cluster matches

veprbl commented on src/algorithms/particle_flow/TrackClusterSubtractor.cc:
```diff
@@ -0,0 +1,244 @@
+      link.setTo(project.getTrack());
+      link.setWeight(1.0); // FIXME placeholder
+      trace("Matched expected cluster {} to track {}", expect_cluster.getObjectID().index,
+            project.getTrack().getObjectID().index);
```
```suggestion
```

veprbl replied:
duplicated trace

ruse-traveler replied:
Whoops! Good catch!

ruse-traveler committed:
Apply suggestion from @veprbl

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

veprbl reviewed (approved):
LGTM

veprbl committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one
